File: sources/src/main/javascript/graph/depth-first-search.js
"use strict";

/**
 * 방향 그래프에서 출발 정점 기준 DFS 전위 순회.
 * Domain public API for `depth-first-search`. JSON 계약 어댑터는 bundles/runner 에 둔다.
 *
 * 이웃 순서는 간선 입력 순서. 도달 불가능한 정점은 결과에 포함하지 않는다.
 * 재귀 깊이 한계를 피하기 위해 명시적 스택으로 재귀 전위 순회를 모사한다.
 *
 * @param {string[]} vertices
 * @param {{from: string, to: string}[]} edges
 * @param {string} source
 * @returns {{order: string[]}}
 */
function execute(vertices, edges, source) {
  if (!Array.isArray(vertices) || vertices.some(vertex => typeof vertex !== "string")) {
    const error = new Error("vertices must be an array of strings");
    error.code = "INVALID_INPUT";
    throw error;
  }
  const adjacency = new Map();
  for (const vertex of vertices) {
    if (adjacency.has(vertex)) {
      const error = new Error("vertex IDs must be unique");
      error.code = "INVALID_INPUT";
      throw error;
    }
    adjacency.set(vertex, []);
  }
  if (typeof source !== "string" || !adjacency.has(source)) {
    const error = new Error("source must be one of vertices");
    error.code = "INVALID_SOURCE";
    throw error;
  }
  if (!Array.isArray(edges)) {
    const error = new Error("edges must be an array");
    error.code = "INVALID_EDGE";
    throw error;
  }
  for (const edge of edges) {
    if (
      edge === null
      || typeof edge !== "object"
      || typeof edge.from !== "string"
      || typeof edge.to !== "string"
      || !adjacency.has(edge.from)
      || !adjacency.has(edge.to)
    ) {
      const error = new Error("edge endpoint must be in vertices");
      error.code = "INVALID_EDGE";
      throw error;
    }
    adjacency.get(edge.from).push(edge.to);
  }

  const visited = new Set([source]);
  const order = [source];
  // frame.next: 아직 시도하지 않은 이웃 인덱스
  const stack = [{ vertex: source, next: 0 }];
  while (stack.length > 0) {
    const frame = stack[stack.length - 1];
    const neighbors = adjacency.get(frame.vertex);
    if (frame.next >= neighbors.length) {
      stack.pop();
      continue;
    }
    const next = neighbors[frame.next++];
    if (visited.has(next)) continue;
    visited.add(next);
    order.push(next);
    stack.push({ vertex: next, next: 0 });
  }
  return { order };
}

module.exports = { execute };

File: sources/src/main/csharp/Graph/DepthFirstSearch.cs
namespace Ssw.Archive.Graph;

/// <summary>Traverses the source-reachable directed graph in depth-first preorder.</summary>
/// <remarks>
/// Vertex IDs are ordered unique strings, source and every edge endpoint are declared, and outgoing neighbors follow
/// edge input order. The iterative frame stack exactly models recursive preorder while supporting deep graphs;
/// unreachable vertices are omitted. Inputs are not mutated.
/// source에서 시작해 현재 정점의 첫 미방문 이웃을 끝까지 따라간 뒤 stack을 되감아 다른 branch를 탐색해요.
/// 정점은 stack에 넣을 때 방문 처리되어 preorder 결과에 정확히 한 번만 등장하는 불변식을 유지해요.
/// 인접 간선 입력 순서가 branch 선택을 결정하며 중복 ID·누락 endpoint·source 오류는 탐색 전에 거부해요.
/// 시간 복잡도는 O(V + E), 방문 집합과 frame stack 공간 복잡도는 O(V)예요.
/// </remarks>
public static class DepthFirstSearch {
    /// <summary>Represents one directed edge.</summary>
    public sealed record Edge(string From, string To);

    /// <summary>Returns source-reachable vertices in deterministic depth-first preorder.</summary>
    public static IReadOnlyList<string> Execute(
        IReadOnlyList<string> vertices,
        IReadOnlyList<Edge> edges,
        string source) {
        ArgumentNullException.ThrowIfNull(vertices);
        ArgumentNullException.ThrowIfNull(edges);
        var adjacency = new Dictionary<string, List<string>>(StringComparer.Ordinal);
        foreach (var vertex in vertices) {
            ArgumentNullException.ThrowIfNull(vertex);
            if (!adjacency.TryAdd(vertex, [])) throw new ArgumentException("Vertex IDs must be unique.", nameof(vertices));
        }
        foreach (var edge in edges) {
            ArgumentNullException.ThrowIfNull(edge);
            if (!adjacency.ContainsKey(edge.From) || !adjacency.ContainsKey(edge.To)) {
                throw new ArgumentException("Edge endpoint is not declared.", nameof(edges));
            }
            adjacency[edge.From].Add(edge.To);
        }
        ArgumentNullException.ThrowIfNull(source);
        if (!adjacency.ContainsKey(source)) throw new ArgumentException("Source is not declared.", nameof(source));
        var visited = new HashSet<string>(StringComparer.Ordinal) { source };
        var order = new List<string> { source };
        var stack = new List<Frame> { new(source) };
        while (stack.Count > 0) {
            var frame = stack[^1];
            var neighbors = adjacency[frame.Vertex];
            if (frame.Next >= neighbors.Count) {
                stack.RemoveAt(stack.Count - 1);
                continue;
            }
            var next = neighbors[frame.Next++];
            if (!visited.Add(next)) continue;
            order.Add(next);
            stack.Add(new Frame(next));
        }
        return order;
    }

    private sealed class Frame(string vertex) {
        internal string Vertex { get; } = vertex;
        internal int Next { get; set; }
    }
}
